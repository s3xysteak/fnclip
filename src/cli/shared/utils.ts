import type { PackageJson } from 'pkg-types'
import type { FnclipOptions } from './options'
import * as pkg from 'empathic/package'
import fs from 'fs-extra'
import * as path from 'pathe'
import { fnclipPath } from '../..'

export async function getMeta() {
  const res: Record<string, string> = await fs.readJson(path.join(fnclipPath, 'funcs-meta.json'))
  return res
}

export const exportContent = (f: string) => `export * from './${f}';\n`

export function ensureExt(fullname: string, ext: string) {
  return path.extname(fullname) ? fullname : `${fullname}${ext}`
}

export async function isTypescript(cwd: string) {
  const packageJsonPath = pkg.up({ cwd })
  if (!packageJsonPath)
    return

  const obj: PackageJson = await fs.readJson(packageJsonPath)
  return !!(obj.dependencies?.typescript || obj.devDependencies?.typescript)
}

export async function updateIndex(opts: FnclipOptions) {
  if (!opts.index)
    return

  const dirPath = path.join(opts.cwd, opts.dir)
  const indexPathMaybeExt = path.join(dirPath, opts.indexPath)

  // check exist index file
  let indexRealPath: string
  for (const ext of ['.ts', '.js']) {
    const targetPath = ensureExt(indexPathMaybeExt, ext)
    if (await fs.exists(targetPath)) {
      indexRealPath = targetPath
      break
    }
  }

  indexRealPath ??= ensureExt(indexPathMaybeExt, (opts.ts ? '.ts' : '.js'))
  await fs.ensureFile(indexRealPath)

  const funcs = await fs.readdir(dirPath)
  await fs.writeFile(indexRealPath, addIgnoreToContent(funcs.map(exportContent).join('')))
}

export function addIgnoreToContent(content: string) {
  const comments = `/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck

`
  return content.includes(comments) ? content : comments + content
}
