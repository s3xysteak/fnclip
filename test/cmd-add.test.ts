import * as fs from 'node:fs/promises'
import { afterAll, describe, expect, it } from 'vitest'
import { exists, rootJoin, run } from './shared'

describe('add', () => {
  const BASE = 'test-add'
  const join = rootJoin.createJoin(BASE)

  afterAll(() => fs.rm(join(), { recursive: true }))

  it('should work', async () => {
    await run(`pnpm fnclip add get --dir ${BASE}`)
    expect(await exists(join('get.js'))).toBe(true)
    expect(await exists(join('index.js'))).toBe(true)

    expect(await fs.readFile(join('get.js'), 'utf8')).toMatchSnapshot()
  })
})
