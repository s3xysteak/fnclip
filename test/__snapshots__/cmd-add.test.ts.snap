// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add > should work 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// biome-ignore lint: disable
// Generated by fnclip

//#region src/functions/object/get/index.ts
/**
* @example
*
* \`\`\`js
* get({ friends: [{ name: 'me' }] }, 'friends[0].name') // -> 'me'
* get([{hello: 'world'}], '[0].hello') // -> 'world'
* \`\`\`
*
* @see {@link https://s3xysteak.github.io/fnclip/functions/object/get/}
*/
function get(value, path, defaultValue) {
	const segments = path.split(/[.[\\]]/g);
	let current = value;
	for (const key of segments) {
		if (current === null) return defaultValue;
		if (current === void 0) return defaultValue;
		const deQuoted = key.replace(/['"]/g, "");
		if (deQuoted.trim() === "") continue;
		current = current[deQuoted];
	}
	if (current === void 0) return defaultValue;
	return current;
}

//#endregion
export { get };"
`;
